src/OpenTelemetry.Extensions/Trace/AutoFlushActivityProcessor.cs

```diff
@@ -0,0 +1,111 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+
+namespace OpenTelemetry.Trace
+{
+    /// <summary>
+    /// Activity processor that flushes its containing <see cref="TracerProvider"/> if an ended
+    /// activity matches a predicate.
+    /// </summary>
+    /// <remarks>
+    /// Add this processor *after* exporter related span processors.
+    /// </remarks>
+    public sealed class AutoFlushActivityProcessor : BaseProcessor<Activity>
+    {
+        private readonly Predicate<Activity> predicate;
+        private readonly int timeoutMilliseconds;
+
+        private TracerProvider tracerProvider;
+        private bool canForceFlush = true;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AutoFlushActivityProcessor"/> class.
+        /// </summary>
+        /// <param name="predicate">Predicate that should return <c>true</c> to initiate a flush.</param>
+        /// <param name="timeoutMilliseconds">Timeout (in milliseconds) to use for flushing. Specify <see cref="Timeout.Infinite"/>
+        /// to wait indefinitely.
+        /// </param>
+        /// <exception cref="ArgumentOutOfRangeException">
+        /// Thrown when the <c>timeoutMilliseconds</c> is smaller than -1.
+        /// </exception>
+        /// <remarks>
+        /// It's assumed that the predicate is defined as a lambda expression which is executed quite fast and
+        /// doesn't contain more complex code. The predicate must not create new Activity instances,
+        /// otherwise the behavior is undefined. Any exception thrown by the predicate will be swallowed and logged.
+        /// In case of an exception the predicated is treated as false which means flush will not be applied.
+        /// </remarks>
+        public AutoFlushActivityProcessor(Predicate<Activity> predicate, int timeoutMilliseconds = 10000)
+        {
+            this.predicate = predicate ?? throw new ArgumentNullException(nameof(predicate));
+            if (timeoutMilliseconds < Timeout.Infinite)
+            {
+                throw new ArgumentOutOfRangeException(nameof(timeoutMilliseconds));
+            }
+
+            this.timeoutMilliseconds = timeoutMilliseconds;
+        }
+
+        /// <inheritdoc/>
+        public override void OnEnd(Activity data)
+        {
+            if (!this.canForceFlush)
+            {
+                return;
+            }
+
+            if (this.tracerProvider == null && this.ParentProvider != null)
+            {
+                this.tracerProvider = this.ParentProvider as TracerProvider;
+                this.canForceFlush = this.tracerProvider != null;
```

Comment:
How about using `this.canForceFlush = this.ParentProvider is TracerProvider;` instead?